cmd/local-artifact-mirror/pull.go

```diff
@@ -0,0 +1,213 @@
+		}
+		return nil
+	},
+	Action: func(c *cli.Context) error {
+		in, err := fetchAndValidateInstallation(c.Context, c.Args().First())
+		if err != nil {
+			return err
+		}
+
+		from := in.Spec.Artifacts.HelmCharts
+		logrus.Infof("fetching helm charts artifact from %s", from)
+		location, err := pullArtifact(c.Context, from)
+		if err != nil {
+			return fmt.Errorf("unable to fetch artifact: %w", err)
+		}
+		defer func() {
+			logrus.Infof("removing temporary directory %s", location)
+			os.RemoveAll(location)
+		}()
+
+		dst := defaults.EmbeddedClusterChartsSubDir()
+		src := filepath.Join(location, HelmChartsArtifactName)
+		logrus.Infof("uncompressing %s", src)
+		if err := tgzutils.Uncompress(src, dst); err != nil {
+			return fmt.Errorf("unable to uncompress images: %w", err)
+		}
+
+		logrus.Infof("helm charts materialized under %s", dst)
+		return nil
+	},
+}
+
+// binariesCommands pulls the binary artifact from the registry running in the cluster and stores
+// it locally. This command is used during cluster upgrades when we want to fetch the most up to
+// date binaries. The binaries is stored in the /usr/local/bin directory and they overwrite the
+// existing binaries.
+var binariesCommand = &cli.Command{
+	Name:      "binaries",
+	Usage:     "Pull binaries artifacts for an airgap installation",
+	UsageText: `embedded-cluster-operator pull binaries <installation-name>`,
+	Before: func(c *cli.Context) error {
+		if os.Getuid() != 0 {
+			return fmt.Errorf("pull binaries command must be run as root")
+		}
+		if len(c.Args().Slice()) != 1 {
+			return fmt.Errorf("expected installation name as argument")
+		}
+		return nil
+	},
+	Action: func(c *cli.Context) error {
+		in, err := fetchAndValidateInstallation(c.Context, c.Args().First())
+		if err != nil {
+			return err
+		}
+
+		from := in.Spec.Artifacts.EmbeddedClusterBinary
+		logrus.Infof("fetching embedded cluster binary artifact from %s", from)
+		location, err := pullArtifact(c.Context, from)
+		if err != nil {
+			return fmt.Errorf("unable to fetch artifact: %w", err)
+		}
+		defer func() {
+			logrus.Infof("removing temporary directory %s", location)
+			os.RemoveAll(location)
+		}()
+
+		bin := filepath.Join(location, EmbeddedClusterBinaryArtifactName)
+		if err := os.Chmod(bin, 0755); err != nil {
+			return fmt.Errorf("unable to change permissions on %s: %w", bin, err)
+		}
+
+		out := bytes.NewBuffer(nil)
+		cmd := exec.Command(bin, "materialize")
+		cmd.Stdout = out
+		cmd.Stderr = out
+
+		logrus.Infof("running command: %s materialize", bin)
+		if err := cmd.Run(); err != nil {
+			logrus.Error(out.String())
+			return err
+		}
+
+		logrus.Infof("embedded cluster binaries materialized")
+		return nil
+	},
+}
+
+// fetchAndValidateInstallation fetches an Installation object from its name and
+// checks if it is valid for an airgap cluster deployment.
+func fetchAndValidateInstallation(ctx context.Context, iname string) (*v1beta1.Installation, error) {
```

Comment:
should we also validate that the necessary artifacts are provided? that shouldn't happen, but we decided to defer the validation to EC in https://github.com/replicatedhq/kots/pull/4524#discussion_r1539950845

Reply:
the fields are [required](https://github.com/replicatedhq/embedded-cluster-operator/blob/main/charts/embedded-cluster-operator/charts/crds/templates/resources.yaml#L238-L242) so they can't be empty. if that is what you mean.